Customize your store header

samueljseay opened the pull request:
### Changes


This PR closes #39718 by loading header patterns in the header customization menu of the assembler.

I've also made some structural changes to move the `EntityProvider` higher up into the layout. I've done this because it will allow the side navigation to share access to the blocks being previewed in `BlockPreview`. I also added a helper hook `useEditorBlocks` so that it's simple for other components to gain access to the blocks and manipulate them.


### Preview

https://github.com/woocommerce/woocommerce/assets/1281828/8b265380-eb98-406a-9c05-76b14c324715




### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

Add header customization to the Assembler Hub

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

samueljseay committed:
Fetch pattern list and filter by headers to display in side-nav.

samueljseay committed:
Move the block context up, share blocks of currently edited post via a hook.

samueljseay committed:
Remove redundant change to a unrelated project.

samueljseay commented on plugins/woocommerce-admin/webpack.config.js:
```diff
@@ -169,7 +169,7 @@ const webpackConfig = {
 		...styleConfig.plugins,
 		// Runs TypeScript type checker on a separate process.
-		! process.env.STORYBOOK && new ForkTsCheckerWebpackPlugin(),
+		// ! process.env.STORYBOOK && new ForkTsCheckerWebpackPlugin(),
```
Ignore, just me getting frustrated by TS errors while trying to develop :P I'll remove this when it goes out of draft.

samueljseay committed:
Update lock file.

samueljseay committed:
Remove unused imports and commented code.

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into dev/choose-header-pattern

samueljseay committed:
Rename useEditBlocks to useEditorBlocks, ignore TS errors.

samueljseay commented on plugins/woocommerce-admin/client/customize-store/assembler-hub/layout.tsx:
```diff
@@ -66,119 +68,136 @@ export const Layout = () => {
-				<motion.div
-					className="edit-site-layout__header-container"
-					animate={ 'view' }
+			<EntityProvider kind="root" type="site">
```
This diff is hard to read but all I did was move EntityProvider here so we can access blocks from inside side nav

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

samueljseay committed:
Ignore typing issues, reenable type checking in webpack.

chihsuan commented on plugins/woocommerce-admin/client/customize-store/assembler-hub/sidebar/sidebar-navigation-screen-header.tsx:
```diff
@@ -30,9 +66,30 @@ export const SidebarNavigationScreenHeader = () => {
 		/>
 	);
+
+	// return null;
```
```suggestion

```

chihsuan reviewed (changes requested):
Hey @samueljseay, Thanks for adding this super quickly. 👍 

I just noticed some issues:

1. It seems that the header patterns cannot be selected again when reloading the page after saving. Could you take a look?
https://github.com/woocommerce/woocommerce/assets/4344253/d7ad4f42-57d9-476e-bf82-6a445351c4d2

2. We should add a focus style border to the selected header pattern on the sidebar. Also need to start with the opacity preview frame (except the header) — when the user scrolls down, bring it back to 100%.

<img width="1280" alt="Header" src="https://github.com/woocommerce/woocommerce/assets/4344253/2bf6cabd-5d88-4d5c-864d-e0f5457b0cdc">

3. We only need to show these four header patterns (Sorry, didn't have it in the issue):

Slug:
- woocommerce-blocks/header-centered-menu-with-search
- woocommerce-blocks/header-essential
- woocommerce-blocks/header-large
- woocommerce-blocks/header-minimal

moon0326 reviewed (commented):
Thank you for the changes :) This is a good reference PR for me.

While testing the changes, I've noticed a possible bug in an edge case.

Steps to reproduce

1. Go to the header page
2. Refresh
3. Click `Assembler Hub`
4. Notice the blocks render twice.

https://github.com/woocommerce/woocommerce/assets/4723145/41c538eb-d101-4719-a717-a7c02fe12217

samueljseay committed:
Filter to the woo header patterns, always replace first block when selecting header.

samueljseay committed:
Remove first block.

samueljseay commented:
@moon0326 I explored the flickering. As you mentioned it happens if you page refresh and it auto navigates to this screen. unfortunately I think this is some bug with the rendering of the layout and navigation screen code that causes redundant re-renders of `SidebarNavigationScreenHeader`, which in turn re-renders the `BlockPatternList`. I haven't got to the bottom of this yet, but I'm thinking maybe it needs to be solved in a separate PR.

samueljseay committed:
Move hooks into assembler folder, add context for highlighted block.

samueljseay committed:
Improve opacity setting code.

samueljseay commented:
@chihsuan I've fixed the issue with saving, I've added some code to handle opacity too, but do you have a suggestion for overriding styles for items in `BlockPatternList`? kinda feels like we need something like emotion to add some targeted styles, otherwise I can add some global styles to fix it.

chihsuan reviewed (approved):
> I've fixed the issue with saving, I've added some code to handle opacity too, but do you have a suggestion for overriding styles for items in BlockPatternList? kinda feels like we need something like emotion to add some targeted styles, otherwise I can add some global styles to fix it.

Hey @samueljseay Thanks for the fix! I confirm that I can save changes successfully 👍  I think adding global styles is probably the only way at the moment. We could file an issue on Gutenberg repo to support that. 😄 

There seem to be a few issues but I think we can address as a follow-up to unblock other tasks as discussed:

1. After jumping out of the header sidebar menu, opacity still exists. We should remove it. 🙂 

https://github.com/woocommerce/woocommerce/assets/4344253/5f7a83dc-3f84-4719-945e-2576ee212cf4

2. I haven't been able to reliably reproduce this issue yet, but it seems that there are some things that may cause the page to become unresponsive. 🤔 

![Screenshot 2023-09-14 at 10 27 25](https://github.com/woocommerce/woocommerce/assets/4344253/3cb6b9c4-536c-4406-9cb9-b677e904760b)

samueljseay committed:
Reset highlighted block when navigating back from header customization.

samueljseay commented:
> After jumping out of the header sidebar menu, opacity still exists. We should remove it. 🙂

😅 good point, This wasn't hard so I added it.

> I haven't been able to reliably reproduce this issue yet, but it seems that there are some things that may cause the page to become unresponsive. 🤔

Ok I haven't seen this yet but I'll keep an eye on it 👁️

samueljseay committed:
Remove unused import.

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into dev/choose-header-pattern
